Font Library: Frontend [Stage 1]

matiasbenedetto opened the pull request:
## What?
**Implementation of the frontend for [Font Library](https://github.com/WordPress/gutenberg/issues/54169)**
**[Stage 1](https://github.com/WordPress/gutenberg/issues/52698): Font management and local font uploads.**


## Why?
The [Font Library](https://github.com/WordPress/gutenberg/issues/54169) consists of a font manager for WordPress. The Font Library is available globally, independently of the theme activated, similar to the Media Library. This list of installed fonts and their assets are site-wide available and the users can select the fonts activated (available in the editor) for each theme.


## Testing Instructions
Font Library is enabled by default in this PR.
To disable Font Library, add this line `define( 'FONT_LIBRARY_DISABLE', true );` to your `wp-config.php`.

~~To make the fonts library backend work you need to add this line in your wp-config.php file:
`define( 'FONT_LIBRARY_ENABLE', true );`~~


## :test_tube:  Test in WordPress Playground:
You can try this PR on [WordPress Playground following this link](https://playground.wordpress.net/#{%22landingPage%22:%20%22/wp-admin/site-editor.php?canvas=edit%22,%20%22steps%22:%20[{%22step%22:%20%22defineWpConfigConsts%22,%20%22consts%22:%20{%22WP_DEVELOPMENT_MODE%22:%20%22all%22}%20},%20{%22step%22:%20%22login%22,%20%22username%22:%20%22admin%22,%20%22password%22:%20%22password%22},%20{%22step%22:%20%22mkdir%22,%20%22path%22:%20%22/wordpress/pr%22},%20{%22step%22:%20%22writeFile%22,%20%22path%22:%20%22/wordpress/pr/pr.zip%22,%20%22data%22:%20{%22resource%22:%20%22url%22,%20%22url%22:%20%22/plugin-proxy?org=WordPress&repo=gutenberg&workflow=Build%20Gutenberg%20Plugin%20Zip&artifact=gutenberg-plugin&pr=53884%22,%20%22caption%22:%20%22Downloading%20Gutenberg%20PR%2053884%22},%20%22progress%22:%20{%22weight%22:%202,%20%22caption%22:%20%22Applying%20Gutenberg%20PR%2053884%22}%20},%20{%22step%22:%20%22unzip%22,%20%22zipPath%22:%20%22/wordpress/pr/pr.zip%22,%20%22extractToPath%22:%20%22/wordpress/pr%22},%20{%22step%22:%20%22installPlugin%22,%20%22pluginZipFile%22:%20{%22resource%22:%20%22vfs%22,%20%22path%22:%20%22/wordpress/pr/gutenberg.zip%22}%20}]%20})


## Disclaimer
We are relying on the `lib-font` package to read the local font assets metadata. Currently, this PR bundles the source code of that library on the source code. We may find a better alternative for that: https://github.com/WordPress/gutenberg/issues/53653

## This PR depends on: 
- [x] https://github.com/WordPress/gutenberg/pull/54161
- [x] https://github.com/WordPress/gutenberg/pull/54167

## Follow up tasks:
- [ ] https://github.com/WordPress/gutenberg/issues/53653

## More info:
Tracking issue: https://github.com/WordPress/gutenberg/issues/52698

matiasbenedetto committed:
adding files from original branch try/font-library

matiasbenedetto committed:
importing remaining files

matiasbenedetto committed:
renaming font library to fonts library

matiasbenedetto committed:
renaming FontLibraryModal to FontsLibraryModal

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend

matiasbenedetto committed:
update frontend based on backend changes

matiasbenedetto committed:
TEMPORAL FIX: Adding Google fonts endpoint a font families list

matiasbenedetto committed:
update uninstall functionality to latest backend implementation

matiasbenedetto committed:
fix loadedUrls state

matiasbenedetto committed:
fixing library not refreshing after uninstall

matiasbenedetto committed:
move uninstall notification functionality to context

matiasbenedetto committed:
Add notification on snackbar when installing fonts

matiasbenedetto committed:
lint js

matiasbenedetto committed:
Merge branch trunk into try/fonts-library-frontend

matiasbenedetto committed:
merge with trunk

matiasbenedetto committed:
renaming FontS Library to Font Library

matiasbenedetto committed:
Removing google_fonts api endpoint

matiasbenedetto committed:
Removing Google fonts specific code

matiasbenedetto committed:
refresh library after local fonts upload

matiasbenedetto committed:
remove unwanted console.log

matiasbenedetto committed:
css fix

matiasbenedetto committed:
removing file key from fontface data sent to the server

matiasbenedetto committed:
open local fonts tab when click the + button

matiasbenedetto committed:
remove font slug prefix

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-rest-font-library-controller.php:
```diff
@@ -67,6 +67,7 @@ public function register_routes() {
 				),
 			)
 		);
+
```
Empty line is not necessary and would be inconsistent with Core.
```suggestion
```

matiasbenedetto committed:
removing empty line

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

matiasbenedetto committed:
check by source (theme and custom)if fonts are activated

matiasbenedetto committed:
delete demo related component and function

matiasbenedetto committed:
remove font family previews

matiasbenedetto committed:
modal width

matiasbenedetto committed:
implementing mockups look for library font variants

matiasbenedetto committed:
reduce margin

matiasbenedetto committed:
display unintall button disabled when the font can not be uninstalled

matiasbenedetto committed:
removing preview controls link

matiasbenedetto committed:
removing unused imports

matiasbenedetto commented:
I changed the previous UI implementation to match the UI mockups defined in https://github.com/WordPress/gutenberg/issues/45271

Here is a screencast of the current state of the UI.
I would appreciate if you could try this PR and provide feedback about the enhancements and fixes needed. 
It has little differences with the mockups mainly around things that were not planned in the mockups or things I think are better how they are now as the local fonts upload tab.


https://github.com/WordPress/gutenberg/assets/1310626/8130f755-a9d4-4f8f-9216-4ff249e0c034

matiasbenedetto committed:
lint css

matiasbenedetto committed:
lint css

matiasbenedetto committed:
fix import

jasmussen commented:
Nice work, took this for a spin. I'm going to assume that Google Fonts integration is being separated out, that makes sense enough to me. 

I also think it's important we start to land these PRs, so I'll review in terms of getting approvals for these in short order, even if we might need to revisit some of the details down the road. 

First off a quick GIF showing entering global styles > typography, then clicking the Plus button to "add font", then clicking through the two tabs for manage and upload:

![fonts](https://github.com/WordPress/gutenberg/assets/1204802/6b491a6a-3296-473a-ac03-25b510e6f9ed)

In general I'm skeptical that we need a separate upload tab, because ideally the whole window can be an upload dropzone. But if it gets us unblocked in terms of the Google Fonts and custom fonts aspect, let's keep it for the time being and then go from there. 

Here's then how uploading a font looks, where I'm drag/dropping a variable width font into the upload tab:

![uploading](https://github.com/WordPress/gutenberg/assets/1204802/bebfeda7-4875-4475-b6d2-1fcde8c2418f)

Note here that the font I uploaded, a "Wave font", doesn't actually appear in the library tab immediatly upon upload. A snackbar also fires, but that snackbar is _below the modal, below the scrim_. To that end, I don't know that the snackbar is the right notice here. Can we show a [progressbar](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-progressbar--docs)? Even an indeterminate one? 

We should also, upon successful upload, redirect to the new font page, i.e. in this case "Library > Wave font", because the next step you have to perform to get the font working, is to "activate it". Shown here in this GIF:

![activating](https://github.com/WordPress/gutenberg/assets/1204802/c9df7bce-af92-41cb-9f78-1077eb222444)

Actually activating, in this branch, works a bit better than it did in some of the previous branches where it was a bit unclear how it worked when going through the Google Fonts flow. But in general the Library tab feels fairly clear in its current appearance now, denoting number of active styles, so let's try this flow and revisit if need be when Google Fonts arrives.

So to summarize, of all the above feedback, the main thing we need to do in order to land this PR is to ensure that upon successful font upload, we redirect in the modal to the new font. The rest is something we can explore in followsups.

---

There are a number of details, however, which we can fix in this PR, or separately, whichever makes the more sense. The global styles inspector font list looks good:

<img width="289" alt="font list" src="https://github.com/WordPress/gutenberg/assets/1204802/2bc121c1-ce2b-471d-ac05-bb9a44d01bc7">

But let's swap the plus and the Aa icons here, so the plus is to the right, matching the tabs in the modal:

<img width="116" alt="plus and Aa icon" src="https://github.com/WordPress/gutenberg/assets/1204802/690b3be4-6b9e-45b3-8ef8-d9b2bd20c60f">

Inside the modal, it wasn't clear to me how this "Delete permanently" button worked:

<img width="235" alt="delete permanently" src="https://github.com/WordPress/gutenberg/assets/1204802/d1dd8f9c-4a2a-42bb-8a5f-a742b30ac9ac">

Upon investigating it became clear that it only becomes enabled when I've first unchecked every font variant first:
<img width="1381" alt="list of font variants" src="https://github.com/WordPress/gutenberg/assets/1204802/a59765ee-d539-48ed-b73c-8c8192e68b27">

Let's permanently enable it. Why do I have to disable the variants first? Presumably disabling the variants is implied by disabling it. We can put it in the footer, make it a default blue tertiary variant without the underline, like this mockup:

<img width="1440" alt="individual font with delete button" src="https://github.com/WordPress/gutenberg/assets/1204802/35e26705-14ea-4e58-9198-0f30329f504e">

Small detail, it's unclear why this "back" button has a different hover style than the one in the Global Styles inspector, they should be the same:

<img width="191" alt="modal back" src="https://github.com/WordPress/gutenberg/assets/1204802/9834caf0-35ed-4e4e-9efb-66b85354e70b">

<img width="137" alt="global styles back" src="https://github.com/WordPress/gutenberg/assets/1204802/fbd10a1f-6074-48fd-b05c-8ab63b526a48">

Another small detail, the tab arrow has the gray bottom border overlaying the blue active-indicator. Modal:
<img width="241" alt="modal tab" src="https://github.com/WordPress/gutenberg/assets/1204802/0a87fe65-77d1-4496-bbf4-8699cb2d0268">
<img width="340" alt="inspector tab" src="https://github.com/WordPress/gutenberg/assets/1204802/df7fe289-8ec8-4a22-a1f6-3b23555478a5">

It isn't clear to me whether that is an issue with [the core component](https://wordpress.github.io/gutenberg/?path=/docs/components-tabpanel--docs), but something for the components group to address there if it is.

This text:
<img width="346" alt="Drag and drop or select font files to install" src="https://github.com/WordPress/gutenberg/assets/1204802/3d9f6c94-a413-4c69-b515-b649067e2eab">

It needs a colon `:` or a period at the end.

The dropzone:

<img width="807" alt="Screenshot 2023-08-30 at 11 54 19" src="https://github.com/WordPress/gutenberg/assets/1204802/cec1b64b-92fa-4f84-8ba5-41b1067569ea">

[The core dropzone](https://wordpress.github.io/gutenberg/?path=/docs/components-dropzone--docs) is not that great looking at the moment, but I don't think we should diverge here. For example, this is also in use for Featured Image:

<img width="293" alt="featured image panel" src="https://github.com/WordPress/gutenberg/assets/1204802/2565431f-60d7-491e-beb4-a9eabc95db0b">

The dashed line, its radus, and its color, are not used elsewhere in the interface. To that end, let's stick with the existing componentry, mocked up here:
<img width="5760" alt="Upload font files" src="https://github.com/WordPress/gutenberg/assets/1204802/e163b72d-f655-45ff-857e-c8aadb7f8b19">

Can we skip the "confirm upload" screen? This one:

<img width="575" alt="confirm upload" src="https://github.com/WordPress/gutenberg/assets/1204802/328dc261-5ad4-46eb-b239-09ec96362c0b">

When you drag and drop an image onto the featured image zone, or into the media library, it immediately uploads. The behavior should ideally be the same here.

Nice work, as noted, most of the detail feedback here isn't something that has to be adressed in this PR, and can happen as followups. Let me know if that helps!

jasmussen commented:
I updated the full flow mockups in #45271, but they should match a combination of this PR and the feedback shared!

matiasbenedetto commented:
> Nice work, took this for a spin. I'm going to assume that Google Fonts integration is being separated out, that makes sense enough to me.

Yes, this PR is for [stage 1 of the Font Library](https://github.com/WordPress/gutenberg/issues/52698). Google fonts are [planned for stage 2](https://github.com/WordPress/gutenberg/issues/53307). Both of them are intended to be ready for 6.4, the library was broken into stages to make it easier to review and merge.

matiasbenedetto committed:
deactivate font families before uninstalling them

matiasbenedetto commented:
> A snackbar also fires, but that snackbar is below the modal, below the scrim.

This a problem on the modal + snackbar components unrelated to the Font Library implementation. There is a ticket to fix that: https://github.com/WordPress/gutenberg/issues/52609
 
 > To that end, I don't know that the snackbar is the right notice here.

We need someplace where to display notifications and the snackbar seems to be the only available on the editor. So ideas around where to display the notifications are welcome.

matiasbenedetto commented:
> We should also, upon successful upload, redirect to the new font page, i.e. in this case "Library > Wave font", because the next step you have to perform to get the font working, is to "activate it".

Ok, apart from moving to the library font listing tab I'll test activating the just uploaded fonts by default. What do you think?

matiasbenedetto committed:
flip buttons order

matiasbenedetto commented:
> But let's swap the plus and the Aa icons here, so the plus is to the right, matching the tabs in the modal.

Done https://github.com/WordPress/gutenberg/pull/53884/commits/7e4a414b31d8e07c370104845fff3c9ee6853b46

matiasbenedetto commented:
> Inside the modal, it wasn't clear to me how this "Delete permanently" button worked:
> Upon investigating it became clear that it only becomes enabled when I've first unchecked every font variant first:
list of font variants.
> Let's permanently enable it. Why do I have to disable the variants first? Presumably disabling the variants is implied by disabling it. 

Done: https://github.com/WordPress/gutenberg/pull/53884/commits/2efb76304962b364a8124f49133d91499035a437

jasmussen commented:
> We need someplace where to display notifications and the snackbar seems to be the only available on the editor. So ideas around where to display the notifications are welcome.

Yes, valid enough, and not a blocker for the PR. But to clarify my question was around that specific notification. IMO uploading a font should be no different than uploading an image, i.e. you drag and drop it on the dropzone, it immediately starts uploading, and you're redirected to the result page. In that flow, there is confirmation in the action, so we don't need additional textual confirmation. (Unless I'm missing nuance here.)

> Ok, apart from moving to the library font listing tab I'll test activating the just uploaded fonts by default. What do you think?

Definitely!

I took the latest for a spin this morning, and I still see the Delete permanently button. As noted, we can do this as a followup, happy to approve this PR without it, but it deserves to always work and be enabled for uploaded fonts, and sit in a more useful place in the footer.

I did notice what may be a bug, OR it might be an issue with my environment, let me know. But shown here in this GIF:

* Click Global Styles > Typography > then click an uploaded font in the listing.
* Dialog opens, cool.
* Close modal, now click "System font"
* Dialog opens. Now click to deactivate the system font. 

![bug maybe](https://github.com/WordPress/gutenberg/assets/1204802/1ff1af3f-f26a-40e4-af01-a1f51b87d353)

As you can maybe glean behind the scrim, the font listing suddenly breaks as you deactivate the system font, and in fact all the uploaded fonts disappear from the list, curiously.

Subsequently when I reloaded the page, I got the following "Cannot redeclare class WP_Font_Face" error, which now prevents me from further testing the PR:

<img width="1392" alt="cannot redeclare class error" src="https://github.com/WordPress/gutenberg/assets/1204802/310be839-0e21-4ce9-9f68-65b73240119d">

In order to keep things nimble, I'm happy to approve this after that issue is fixed (or we figure out that it was an issue on my end). Thanks for the work.

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto committed:
Moving uninstall button to modal footer

matiasbenedetto committed:
be able to indicate nested paths on __experimentalSaveSpecifiedEntityEdits

matiasbenedetto committed:
save only fontFamilies when persisting global styles from font library modal

matiasbenedetto committed:
Add edits data to the useEntityRecord

matiasbenedetto commented:
> Subsequently when I reloaded the page, I got the following "Cannot redeclare class WP_Font_Face" error, which now prevents me from further testing the PR

This was fixed by merging with the latest trunk version.

matiasbenedetto committed:
disable save button when there are no changes to be saved

matiasbenedetto committed:
add notification when font families are updated

matiasbenedetto committed:
upload local fonts just after selecting them

matiasbenedetto committed:
Removing upload local fonts tab from modal

matiasbenedetto committed:
drop zone background color

matiasbenedetto committed:
remove unused import

matiasbenedetto commented:
> Can we skip the "confirm upload" screen?

Yep, done! I implemented the version you proposed and removed the upload tab, because without the confirmation, it makes no sense:

![image](https://github.com/WordPress/gutenberg/assets/1310626/ab5cbb03-7249-4a33-a8cb-fc5e135dce97)

matiasbenedetto committed:
activate fonts just after installed

matiasbenedetto committed:
fix typo

matiasbenedetto committed:
hide confirm dialog after deleting font

matiasbenedetto committed:
remove + button

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto commented:
This is a screencast of the latest version of this PR. It would be great to receive more feedback and testing.

https://github.com/WordPress/gutenberg/assets/1310626/76483afc-4a1d-4f51-8544-35d5833dbd4b

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto committed:
fix merging font families functions

matiasbenedetto committed:
add tests for mergeFontFamilies and mergeFontFaces functions

matiasbenedetto committed:
add tests  for getFontFaceVariantName

matiasbenedetto committed:
fix logic bug and add tests for isUrlEnconded function

matiasbenedetto committed:
fix logic bug in setUIValuesNeeded function

matiasbenedetto committed:
add tests for  setUIValuesNeeded function

ironprogrammer commented on packages/edit-site/src/components/global-styles/font-library-modal/local-fonts.js:
```diff
@@ -0,0 +1,181 @@
+				onChange={ onFilesUpload }
+				render={ ( { openFileDialog } ) => (
+					<div className="font-library-modal__upload-area" onClick={ openFileDialog }>
+						<span>{ __("Drag and drop you font files here.") }</span>
```
```suggestion
						<span>{ __("Drag and drop your font files here.") }</span>
```
Small typo.

Alternatively, it would be nice to indicate that the area can be clicked to activate the upload dialog:
```suggestion
						<span>{ __("Click to upload, or drag and drop your font files here.") }</span>
```

matiasbenedetto committed:
move utils and utils tests

matiasbenedetto committed:
remove import not needed

matiasbenedetto committed:
remove default demo config

ironprogrammer commented on packages/edit-site/src/components/global-styles/font-library-modal/local-fonts.js:
```diff
@@ -0,0 +1,181 @@
+			<Spacer margin={ 2 } />
+			<DropZone onFilesDrop={ handleDropZone } />
+			<FormFileUpload
+				accept="font/*"
```
```suggestion
				accept="font/*,.woff,.woff2"
```
In macOS Safari and Chrome, using `font/*` for the system upload dialog allows .ttf and .otf, but doesn't allow .woff/.woff2. Should they be explicitly included here to account for these types as defined in `ALLOWED_FILE_EXTENSIONS`?

ironprogrammer reviewed (changes requested):
This is looking terrific, @matiasbenedetto! Here are a couple of suggestions.

matiasbenedetto committed:
fix typo

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

matiasbenedetto committed:
update file extensions allowed in file dialog on mac

In macOS Safari and Chrome, using font/* for the system upload dialog allows .ttf and .otf, but doesn't allow .woff/.woff2.

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

ironprogrammer commented on packages/edit-site/src/components/global-styles/font-library-modal/local-fonts.js:
```diff
@@ -0,0 +1,181 @@
+		} );
+		const loadedFace = await newFont.load();
+		document.fonts.add( loadedFace );
+	} 
```
```suggestion
	}
```
Nit-pick extra space.

matiasbenedetto committed:
remove extra space

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

ironprogrammer commented:
## Test Report

### Steps to Test
_(This list may be incomplete -- please share if there are additional things to test 🙏🏻.)_
1. Review the *Styles > Typography* fonts list for active fonts.
2. Click the "Aa" icon to open the font management dialog.
3. Upload font files using drag and drop.
4. Upload font files using the upload dialog.
5. Verify that font assets are stored in `/wp-content/fonts/`.
6. Activate font variants and compare with font list.
7. Deactivate font variants and compare with font list.
8. Delete an uploaded font family.
9. Verify that deleted font assets are removed from `/wp-content/fonts/`.

### Environment
- OS: macOS 13.5.1
- Browser: Safari 16.6, Google Chrome 116.0.5845.179
- PHP: 8.2.10
- WordPress: 6.4-alpha-56267-src
	- `wp-config.php`: `define( 'FONT_LIBRARY_ENABLE', true );`
- Theme: twentytwentythree v1.2
- Active Plugins:
	- gutenberg v16.6.0-rc.1

### Actual Results
**1-2. Fonts List and Management Dialog**
- ✅ List of fonts in *Styles > Typography* reflects those that are active in the font management dialog.
- ✅ "Aa" icon displays the font management dialog. Button can also be selected and activated using keyboard (e.g. <kbd>tab, return</kbd>).

**3-5. Upload**
- ✅ Font files can be uploaded using drag and drop. Types tested: TTF, OTF, WOFF, WOFF2.
	- ✅ Invalid files cannot be uploaded through drag and drop. Types tested: EOT, SVG, TXT.
- ✅ Font files can be uploaded via upload dialog. Types tested: TTF, OTF, WOFF, WOFF2.
- ✅ Uploaded fonts appear in the font management dialog, and are automatically activated.
	- ❌ Uploading a variant for an existing custom font re-activates other deactivated variants. I expected that deactivated variants would remain deactivated.
- ✅ Re-installing an existing font replaces the existing one.
- ⚠️ Font with same name but different type as existing custom font (e.g. `guillon_normal_400.ttf` and `guillon_normal_400.woff2`) is added as a separate variant.
	- The font management dialog shows "1/2 variants active", even though both are checked. **What is expected here?**
- ✅ Font with same name as a theme font (e.g. "DM Sans") is added as a separate custom font.
	- ✅ Custom font overrides the theme's font in FE (only custom font is printed).
	- ❌ In Typography > Fonts list, first found font by name is used to style list. E.g. DM Sans Light when added as custom font is used to style the theme-supplied font. I expected each font in the list to be styled with its matching font file.
		- Clicking the second font switches this around: the font loaded to display the variant dialog styles both fonts in the list.

| On pageload or after viewing custom DM Sans variants | After viewing theme's DM Sans variants |
|:-:|:-:|
| <img width="269" alt="DM Sans Light used in fonts list" src="https://github.com/WordPress/gutenberg/assets/824344/816e811d-e0d1-4f0a-95bf-1573cf4163ee">  | <img width="269" alt="DM Sans Regular used in fonts list" src="https://github.com/WordPress/gutenberg/assets/824344/9c88a108-0cbe-4665-9bf1-065d4f5a9395"> |

- ✅ Uploaded fonts are stored in `/wp-content/fonts/`.

**6-7. Activate/Deactivate**
- ✅ Activated fonts appear in font list, along with number of active variants.
- ✅ Deactivated fonts are removed from font list.
	- ❌ Deactivating System Font also deactivates all custom font variants. This is unexpected behavior. (Also noted in [#issuecomment-1702256035](https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1702256035).)
	- ❌ Reactivating a custom font while System Font is deactivated causes the latter to be reactivated. I expected System Font to remain deactivated.

**8-9. Delete**
- ✅ Deleted fonts are removed from the font management dialog and font list.
- ✅ Deleting a font removes all registered variant files from `/wp-content/fonts/`.

### Additional Notes
Overall, this is looking really tight! There are a few editor font handling issues that cropped up (e.g. Safari only showed me fallbacks in the editor, and Chrome was inconsistent 🤯), but those may be known issues, so I'll do a little digging to file them separately as needed.

mikachan committed:
Apply coding standards

mikachan committed:
Make font cards buttons

mikachan committed:
Fix linter errors

mikachan committed:
Fix max depth error

mikachan committed:
Revert font lib formatting

mikachan committed:
Fix linter errors

mikachan committed:
Undo e2e test linter fixes

Not related to this PR

mikachan committed:
Remove more e2e test changes

matiasbenedetto commented:
Thanks for the review!

> ⚠️ Font with same name but different type as existing custom font (e.g. guillon_normal_400.ttf and guillon_normal_400.woff2) is added as a separate variant. 

I created an issue for this adding more details and the expected behaviour: https://github.com/WordPress/gutenberg/issues/54253

matiasbenedetto committed:
fixing error persisting activation changes

matiasbenedetto commented:
> ❌ Deactivating System Font also deactivates all custom font variants. This is unexpected behavior. (Also noted in https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1702256035.)
    ❌ Reactivating a custom font while System Font is deactivated causes the latter to be reactivated. I expected System Font to remain deactivated.
    
I think this issue  was fixed with this commit: https://github.com/WordPress/gutenberg/pull/53884/commits/5f8e9559b63da27cfab3201706ef2beb09b9c40b

matiasbenedetto committed:
improving font previews

matiasbenedetto committed:
adding unit tests for getPreviewStyle util

matiasbenedetto commented:
> ❌ In Typography > Fonts list, first found font by name is used to style list. E.g. DM Sans Light when added as custom font is used to style the theme-supplied font. I expected each font in the list to be styled with its matching font file.
> Clicking the second font switches this around: the font loaded to display the variant dialog styles both fonts in the list

This was fixed here: https://github.com/WordPress/gutenberg/pull/53884/commits/8745fd41e599dfe0a85f329c64a257631c492831

matiasbenedetto committed:
lint fix

matiasbenedetto committed:
fix accepted file extensions not working properly on firefox

matiasbenedetto committed:
sort font faces by weight

matiasbenedetto committed:
move the code of toggleActivatedFont to a another file to reduce their size and make it easier to read

matiasbenedetto committed:
adding tests for toggleFont

matiasbenedetto committed:
lint

matiasbenedetto committed:
avoid resetting theme font families when activating or deactivating font famiies when install or uninstall

Ren2049 commented:
In the playground link version using Mac Safari I wasn't able to apply a custom uploaded .ttf font for headings (DXS Guildhall) 

It showed up in the font manager and preview thumbnail but the editor didn't pick it up and fell back to Times, which wasn't even in the list of fonts.

I've also noticed that after deletion of the uploaded font and uploading another one, it still showed the deleted font in the preview for a short while.

In the beginning I had to search for the font library feature, took me a while figuring out that I'd have to click the letters next to Fonts. Maybe use a "font library" text label or button instead of an icon?

madhusudhand committed:
fix issue where fontFace length is zero

mikachan committed:
Add deps comment to toggleFont file

mikachan committed:
Fix theme and custom undefined errors

mikachan committed:
Add initial e2e tests

mikachan committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

mikachan committed:
Refactor e2e test order

mikachan commented:
I've added some initial e2e tests to this PR, but I didn't want to hold up this PR or make it much larger than it already is, so I'm planning to continue in this separate PR: https://github.com/WordPress/gutenberg/pull/54298. We can always merge that into this PR if there are not many more changes before this one is merged.

To confirm, I'm happy with the e2e tests added in this PR so far, so they can be safely merged.

matiasbenedetto committed:
avoid re-activating all the font faces of a installed font when a new font face is installed

matiasbenedetto committed:
adding tests for utils functions

matiasbenedetto committed:
renaming parameters

matiasbenedetto committed:
adding jsdoc for function

matiasbenedetto commented:
> In the beginning I had to search for the font library feature, took me a while figuring out that I'd have to click the letters next to Fonts. Maybe use a "font library" text label or button instead of an icon?

@jasmussen you may be interested on this feedback

matiasbenedetto committed:
remove code no longer needed

matiasbenedetto committed:
avoid potential errors with non existing fontFace array

matiasbenedetto committed:
Revert "avoid potential errors with non existing fontFace array"

This reverts commit 43dae169d80290f00486dbbe729e70042c731e8a.

ironprogrammer commented:
Thanks for the updates! The originally flagged items (❌, ⚠️) in [#issuecomment-1709315408](https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1709315408) have been addressed or tracked in a new issue.
## Test Report

### Environment
- OS: macOS 13.5.1
- Browser: Safari 16.6, Google Chrome 116.0.5845.179
- PHP: 8.2.10
- WordPress: 6.3.2-alpha-56487-src
	- `wp-config.php`: `define( 'FONT_LIBRARY_ENABLE', true );`
- Theme: twentytwentythree v1.2
- Active Plugins:
	- gutenberg v16.6.0

### Actual Results
_(These results reflect re-testing the most recent commits.)_
**3-5. Upload**
- ✅ Uploading a variant for an existing custom font does not re-activate other deactivated variants ([14aa39a](https://github.com/WordPress/gutenberg/pull/53884/commits/14aa39ac60bab9a96c891d0ac47eb320f168bd77)).
- ✅ Fonts in the list are styled by their active variants, even if a custom and theme font share the same name ([8745fd4](https://github.com/WordPress/gutenberg/commit/8745fd41e599dfe0a85f329c64a257631c492831)).

| Custom DM Sans families in sidebar | Custom DM Sans families in modal |
|:-:|:-:|
| ![](https://cldup.com/US7262xKmS.png) | ![](https://cldup.com/8hkrfynVEw.png) |

**6-7. Activate/Deactivate**
- ✅ Deactivating System Font does not impact custom fonts ([5f8e955](https://github.com/WordPress/gutenberg/commit/5f8e9559b63da27cfab3201706ef2beb09b9c40b)).
- ✅ Reactivating a custom font while System Font is deactivated does not impact custom fonts (also [5f8e955](https://github.com/WordPress/gutenberg/commit/5f8e9559b63da27cfab3201706ef2beb09b9c40b)).

matiasbenedetto committed:
fix error uploading several font faces from different font families

Commit:
Fix font library modal prop typo

jordesign commented:
Loving this generally - nice work.

* +1 to the Aa icon not being clear it was the action/trigger to manage my fonts (but I see it was noted previously).
* I noticed that when I removed/disabled fonts the confirmation notice/snackbar appears below the modal, so it was hard to spot...

<img width="635" alt="Screenshot 2023-09-11 at 1 40 09 pm" src="https://github.com/WordPress/gutenberg/assets/1842363/45887e42-016b-4dce-b3cb-ba0ee16497e8">

* When uploading a font family - the snackbar was again obscured, and the fact that the upload drop-area kept focus meant that I didn't know my font had been added at the top of the list. It wasn't visible.

<img width="1076" alt="Screenshot 2023-09-11 at 1 44 01 pm" src="https://github.com/WordPress/gutenberg/assets/1842363/4cceb454-c02d-48eb-9510-cc517f7cbbeb">

> It showed up in the font manager and preview thumbnail but the editor didn't pick it up and fell back to Times, which wasn't even in the list of fonts.

I found this to be the case as well. Whilst the font would upload - it was just showing the default serif font.

<img width="1168" alt="Screenshot 2023-09-11 at 1 46 03 pm" src="https://github.com/WordPress/gutenberg/assets/1842363/284fd948-8e13-411a-a520-f31183ba2218">

<img width="586" alt="Screenshot 2023-09-11 at 1 47 29 pm" src="https://github.com/WordPress/gutenberg/assets/1842363/90ad406a-9682-4ed9-8179-3aa376d82c3b">

jasmussen commented:
Really nice work. A quick review of the general modal, the buttons, activating, in this GIF:

![fonts](https://github.com/WordPress/gutenberg/assets/1204802/d4294d7f-2217-4db0-8915-c45d1128b6b6)

Uploading:

![uploading](https://github.com/WordPress/gutenberg/assets/1204802/dfc48fc3-c1cb-4e73-953b-ebe4bfe0de53)

**A few things need fixing for this PR to land:**

1. This button should not be red, and it should not be underlined, and it should just say "Delete":

<img width="364" alt="Screenshot 2023-09-11 at 09 04 16" src="https://github.com/WordPress/gutenberg/assets/1204802/8fbb670a-78ec-4b65-99d2-a36716d7788a">

2. "Activate & Save" could be the right label in the _modal_, but this PR changes the save button across the whole site editor, and it should not. Shown here in the screenshot, the modal button says "Update", and the global save button behind the scrim says "Activate & Save":
<img width="1392" alt="activate and save" src="https://github.com/WordPress/gutenberg/assets/1204802/599f6514-cc7e-45c3-aa18-d4802499e9d0">

We can change this so:

* In the modal, "Update" is replaced by "Activate & Save"
* In the site editor, the previous label is kept as simpy "Save"

Here's the existing save button label:

<img width="382" alt="save button" src="https://github.com/WordPress/gutenberg/assets/1204802/351f4430-d12c-4db5-85b0-20c209b57568">

3. Remove the snackbar

When you upload a font, a snackbar fires. We don't do this for any other upload, and should not do that here either.

<img width="343" alt="obscured snackbar" src="https://github.com/WordPress/gutenberg/assets/1204802/7baa6acc-32f6-4d4b-b990-f2a334f52645">


**The following can be a followup:** 

When you upload a font, it is correctly uploaded instantly on drag and drop, or on select, but you are redirected to the main Library tab. 

You should not be directed to the main list, you should be directed to the font you just uploaded, in this case I uploaded "Roboto Medium", and so I should be directed to the Roboto Medium page:

<img width="1392" alt="font" src="https://github.com/WordPress/gutenberg/assets/1204802/bb9d6d0c-37cd-445b-aa83-12ba00042bd0">

jasmussen reviewed (changes requested):
This PR is fantastic, and is a brief push away from landing. I went into detail in [this comment](https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1713306944), but for clarity, we should merge this when:

1. The delete button is the default color, not underlined, and says "Delete"
2. The main site editor button is reverted back to saying "Save" on entity saving. The modal button can say "Activate & Save" if need be.
3. The snackbar on upload is removed.

jasmussen commented:
I also noted that the Upload tab was removed again. I'm fine with keeping it at the end, or having a tab — [these mockups still include a tab](https://github.com/WordPress/gutenberg/issues/45271).

Ren2049 commented:
@jasmussen "Manage Fonts" is too hidden behind the Aa icon, users won't find it.

jasmussen commented:
I don't agree, but mainlly we don't have a baseline merged that we can test in the plugin yet, so I think first step is to land this so we are able to speak with more confidence. For now, in this interface you can click either on the plus, the manage icon, or any of the fonts listed below, to manage:

<img width="561" alt="Screenshot 2023-09-11 at 10 22 21" src="https://github.com/WordPress/gutenberg/assets/1204802/db15e4e7-86be-4016-9333-e479fd3e5c4b">

matiasbenedetto committed:
activate the right font faces (the installed ones instead of the local definitions).

ironprogrammer commented:
In response to @Ren2049 and @jordesign:
> It showed up in the font manager and preview thumbnail but the editor didn't pick it up and fell back to Times, which wasn't even in the list of fonts.

The display of newly added fonts in the editor should be addressed with https://github.com/WordPress/gutenberg/pull/54313. When I've cherry-picked it into this PR, the issue is resolved.

vcanales committed:
Font library: load font assets on editor canvas (#54334)

* Font library: load font assets on editor canvas

This change adds loaded font faces to the editor canvas iframe so that they can be used immediately after uploading them.

Re: https://github.com/WordPress/gutenberg/issues/51764

* load font families on iframe on activation

* removing addFontFaceToBrowser function and consolidate it with loadFontFaceInBrowser that was doing almost the same

---------

Co-authored-by: Matias Benedetto <matias.benedetto@gmail.com>

matiasbenedetto committed:
removing success snackbar notifications

matiasbenedetto committed:
update  uninstall font family button look

matiasbenedetto committed:
delete any change to save-button component

matiasbenedetto commented:
> You should not be directed to the main list, you should be directed to the font you just uploaded, in this case I uploaded "Roboto Medium", and so I should be directed to the Roboto Medium page

@jasmussen I'm not sure about this suggestion because you can upload multiple files from many families at the same time. In that case which font family should you be redirected to?

matiasbenedetto commented:
> This PR is fantastic, and is a brief push away from landing. I went into detail in https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1713306944, but for clarity, we should merge this when:
> The delete button is the default color, not underlined, and says "Delete"

Done!

 > The main site editor button is reverted back to saying "Save" on entity saving. The modal button can say "Activate & Save" if need be.
 
I reverted any change previously made in this bach for the entity save button. In the modal, we are using an ad-hoc one that only saves changes to `settings.typography.fontFamilies` path of the global styles/settings. @jasmussen What text should it display?
 
> The snackbar on upload is removed.

Done!

matiasbenedetto committed:
use a simple file id for uploaded files

jasmussen commented:
> I'm not sure about this suggestion because you can upload multiple files from many families at the same time. In that case which font family should you be redirected to?

Good point. Can you know whether you can redirect to a family, or whether multiple families were uploaded? If yes, then redirect to the family if only one family was uploaded, and just go to the library otherwise.

jasmussen committed:
Update delete button to be tertiary.

jasmussen reviewed (approved):
Thank you for the changes. Let's merge this!

There are a couple of followups, but please merge this as-is, and we can do those followups separately.

jasmussen commented:
I pushed a small change so the Delete button is tertiary:

<img width="235" alt="Screenshot 2023-09-12 at 09 16 48" src="https://github.com/WordPress/gutenberg/assets/1204802/b57af4e5-356f-4eb4-b0d8-1c88400234b1">

As noted, let's merge this, but we should do a couple of followups. Perhaps the most important one is that we should not skip the multi-entity saving dialog. Note in this GIF how selecting font variants and pressing "Update" also saves the changes site-wide:

![updating](https://github.com/WordPress/gutenberg/assets/1204802/65c299ce-32b1-4204-afb1-0ee3b06c598b)

The previous behavior was correct: the "Update" button in the modal enqueues the changes to be saved by the global save button in the multi entity saving interface. This is important because it gives you a last minute opportunity to change your mind. The only problem previously was that the button labels were wrong.

Secondly, and as discussed, if we can know you only uploaded a single font family, redirect to that family upon successful upload.

In a separate issue, we discussed font stacks. In the modal, we show "System fonts" as well:
<img width="1481" alt="Screenshot 2023-09-12 at 09 11 51" src="https://github.com/WordPress/gutenberg/assets/1204802/7c29b200-2661-4b8a-aa03-b9fccecb1026">

But if you deactivate this one it becomes a bit problematic for the theme. Can we just not show that option? Or can we gray it? It feels like we should have a couple of base font stacks that are always available. Something non-urgent to look into.

The Inter font, for whatever reason, did not load for me. This is probably me having nuked my local environment a couple of times, but if there was some way to validate a font file was present and actually loading, that'd be a nice enhancement:

<img width="1475" alt="Screenshot 2023-09-12 at 09 12 02" src="https://github.com/WordPress/gutenberg/assets/1204802/4a34590c-d6e6-4e74-acc0-ec1418917e36">

matiasbenedetto committed:
fix test

matiasbenedetto committed:
use FONT_LIBRARY_DISABLE constant to disable Font Library and Font Face loading and load Font API instead

matiasbenedetto committed:
move confirm uninstall dialog to a new component

colorful-tones commented:
## Test Report

### Steps to Test
(Mostly following @ironprogrammer 's handy [testing guide](https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1709315408) with different browser. Note: I did not make it through his entire testing suite.)

1. Review the _Styles > Typography_ fonts list for active fonts.
2. Click the "Aa" icon to open the font management dialog.
3. Upload font files using drag and drop.
4. Upload font files using the upload dialog.
5. Verify that font assets are stored in `/wp-content/fonts/`.

### Environment

- OS: macOS 13.5.2
- Browser: Firefox 117.0
- PHP: 8.1.22
- WordPress: 6.3.1
  - ~~wp-config.php: define( 'FONT_LIBRARY_ENABLE', true );~~ not necessary anymore.
- Theme: twentytwentythree v1.2
- Active Plugins:
  - Gutenberg build from this PR

### Actual results

__1-2. Fonts List and Management Dialog__

- ✅ List of fonts in _Styles > Typography_ reflects those that are active in the font management dialog.
- ✅ "Aa" icon displays the font management dialog. Button can also be selected and activated using keyboard (e.g. `tab`, `return`).

__3-5. Upload__

- ✅ Font files can be uploaded using drag and drop. Types tested: TTF, OTF, WOFF, WOFF2.
  - ✅ Invalid files cannot be uploaded through drag and drop. Types tested: EOT, SVG, TXT.
- ✅ Font files can be uploaded via upload dialog. Types tested: TTF, OTF, WOFF, WOFF2.
- ✅ Uploaded fonts appear in the font management dialog, and are automatically activated.
- ⚠️ Font with same name but different type as existing custom font (e.g. `montserrat_normal_100.woff` and `montserrat_normal_100.woff2`) is added as a separate variant. (see screenshot below)

![Screenshot 2023-09-12 at 6 06 07 PM](https://github.com/WordPress/gutenberg/assets/405912/8d9979d0-22ac-494a-b0ff-9f31dc418481)

This was as far as I got in testing before being pulled away. Overall, there is some great stuff here. 👍

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto commented:
Thanks for the review @colorful-tones !

> ⚠️ Font with same name but different type as existing custom font (e.g. montserrat_normal_100.woff and montserrat_normal_100.woff) is added as a separate variant. (see screenshot below)

This [is a known issue](https://github.com/WordPress/gutenberg/issues/54253) and it's being worked here: https://github.com/WordPress/gutenberg/pull/54320#top. It will be fixed soon

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto committed:
adding progress bar while installing fonts

matiasbenedetto committed:
add early return

matiasbenedetto committed:
format php

matiasbenedetto committed:
Merge branch 'trunk' into try/fonts-library-frontend-stage1

matiasbenedetto committed:
removing not needed changes to save hub button

hellofromtonya reviewed (approved):
Approving based on the amount of testing and reviews this PR has received. Though I have not done a thorough JS code review, this PR should be good for merging. It's time to ship it and get more feedback.

jsnajdr commented:
Update: I see there is #53653 now.

Hi @matiasbenedetto :wave: I have some questions about the organization of the code added by this PR:
- the `inflate` and `unbrotli` modules are third party minified code. Could we add them as NPM package dependencies instead? Or was there a good reason to put a copy in the Gutenberg repo?
- The `lib-font` module also looks like a third party code. Where it comes from?
- why is the code in the `edit-site/lib` directory instead of `src`, like all other source code?

I encountered this code when running `npm run format`, to reformat the codebase with Prettier, and seeing some changes I didn't expect, like formatting minified code. I'm wondering whether the code is in a desired long term state, and we should just add the `lib` directory to `.prettierignore`, or whether we should remove it instead in favor of NPM dependencies.

matiasbenedetto commented:
Hi @jsnajdr, the 3 files are coming from `lib-font` package. To make it work in the browser we needed to make small changes. Adding the files to Gutenberg source code is short-term solution and we are working on adding the package in a more elegant way. There is [an issue](https://github.com/WordPress/gutenberg/issues/53653) and [PR ongoing](https://github.com/WordPress/gutenberg/pull/54440).

tinjure20 commented:
I am following @colorful-tones notes [above](https://github.com/WordPress/gutenberg/pull/53884#issuecomment-1716567977) in a similar environment with both Firefox and Chrome browsers, and the Gutenberg nightly active plugin. I ran the test using Playground as well. But I failed to upload fonts. What am I doing wrong?

tinjure20 commented:
**Update**. Because there was no instructions, I was trying to upload the entire downloaded zip folder. But when I uploaded individual TFF file, the file uploads successfully. Thank you.

matiasbenedetto commented:
> Update. Because there was no instructions, I was trying to upload the entire downloaded zip folder. But when I uploaded individual TFF file, the file uploads successfully. Thank you.

@tinjure20 good to know. Thanks for the update! It could be a common mistake, so maybe we should add some kind of visual hint about the need to install font assets and not other types of files.

tinjure20 commented:
> so maybe we should add some kind of visual hint about the need to install font assets and not other types of files

IMHO, many users would benefit from a short warning note below the drag-and-drop area.

autumnfjeld commented:
A couple small things:

Item 1
<img width="113" alt="image" src="https://github.com/WordPress/gutenberg/assets/5665959/fcbee965-5754-45e8-b0be-bc6fa8ddcf88">
Should be "your" not "you".

Item 2
<img width="245" alt="image" src="https://github.com/WordPress/gutenberg/assets/5665959/13308460-dcfb-4b65-87ef-ff9c028339f1">

"too many" is vague.  Will there be a link out for more info? How will someone  know if 5 is too many or if 15 is too many?

jasmussen commented:
> "too many" is vague. Will there be a link out for more info? How will someone know if 5 is too many or if 15 is too many?

Good callout, how many is too many? If you have a great suggestion for phrasing, we can run with it!

On the technical side, this issue should ideally mostly disappear in the future, as the Fonts API matures to a point where only the fonts that are actually in use on a particular page of the site, are loaded. I don't believe that is the case yet, but it should be tracked for a future improvement. I'll defer to @matiasbenedetto on details here.

So for the time being, it would be good to update the copy. Here's a quick attempt:

> Choose font variants. Note that too many could make your site slower, as each extra variant makes the initial download take longer, with about 160Kb for each variant.

I don't actually like the specific size call-out here, to me it puts it way into developer territory, but if we can think of it as temporary until the Fonts API smartly handles it, it might be acceptable? As noted, feel free to suggest an edit!

matiasbenedetto commented:
Hi :wave:  This PR was already merged. Please, If the wording needs to be updated, open a new issue with the details of the change.
